<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Running NGS computing demanding applications, e.g reads mapping and counting, wrapped in docker containers • docker4seq</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Running NGS computing demanding applications, e.g reads mapping and counting, wrapped in docker containers">
<meta property="og:description" content="A collection of functions to execute NGS computing demanding applications, e.g reads mapping and counting, wrapped in docker containers">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">docker4seq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.2.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="articles/docker4seq.html">Get started</a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/skeleton.html">Skeleton tutorial</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="contents col-md-9">
<div id="reproducible-bioinformatics-community" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#reproducible-bioinformatics-community" class="anchor"></a>Reproducible Bioinformatics Community</h1></div>
<p>The aim of Reproducible Bioinformatics project is the creation of easy to use Bioinformatics workflows that fullfill the following roles (<a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003285">Sandve et al. PLoS Comp Biol. 2013</a>):</p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r"><span class="dv">1</span> For Every Result, Keep Track of How It Was Produced
<span class="dv">2</span> Avoid Manual Data Manipulation Steps
<span class="dv">3</span> Archive the Exact Versions of All External Programs Used
<span class="dv">4</span> Version Control All Custom Scripts
<span class="dv">5</span> Record All Intermediate Results, When Possible <span class="cf">in</span> Standardized Formats
<span class="dv">6</span> For Analyses That Include Randomness, Note Underlying Random Seeds
<span class="dv">7</span> Always Store Raw Data behind Plots
<span class="dv">8</span> Generate Hierarchical Analysis Output, Allowing Layers of Increasing Detail to Be Inspected
<span class="dv">9</span> Connect Textual Statements to Underlying Results
<span class="dv">10</span> Provide Public Access to Scripts, Runs, and Results</code></pre></div>
<p>Reproducible Bioinformatics is a non-profit and open-source project.</p>
<p>We are a group of Bioinformaticians interested to simplify the use of bioinformatics tools to Biologists w/wo scripting ability. At the same time we are interested in providing robust and reproducible workflows.</p>
<p>For this reason we have developed the docker4seq package.</p>
<p>At the present time a total of three workflows are available in the stable version of docker4seq package (more info below in the text):</p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r"><span class="op">-</span><span class="st"> </span>RNAseq workflow
<span class="op">-</span><span class="st"> </span>miRNAseq workflow
<span class="op">-</span><span class="st"> </span>ChIPseq workflow</code></pre></div>
<p>Under development are:</p>
<div class="sourceCode"><pre class="sourceCode R"><code class="sourceCode r"><span class="op">-</span><span class="st"> </span>PDX workflow<span class="op">:</span><span class="st"> </span>variants calling <span class="cf">in</span> patient derived <span class="kw">xenograft</span> (PDX) from RNAseq and EXOMEseq data
<span class="op">-</span><span class="st"> </span>Single cell analysis workflow
<span class="op">-</span><span class="st"> </span>Metagenomics workflow</code></pre></div>
<p>All workflows are controlled by a set of R fuctions, part of docker4seq package, and the algorithms used are all encapsulated into Docker images and stored at docker.io/repbioinfo repository.</p>
<p>More info on docker4seq: <a href="https://kendomaniac.github.io/docker4seq/index.html">docker4seq web page</a></p>
<div id="how-to-be-part-of-the-reproducible-bioinformatics-project-community" class="section level3">
<h3 class="hasAnchor">
<a href="#how-to-be-part-of-the-reproducible-bioinformatics-project-community" class="anchor"></a>How to be part of the Reproducible Bioinformatics Project community</h3>
<p>Any bioinformatician interested to embed specific applications in the available workflows or interested to develop a new workflow is requested to embed the application(s) in a docker image, save it in a public repository and configure one or more R functions that can be used to interact with the docker image. The module/workflow needs to fullfil at least the first 6 Sandve’s rules.</p>
<p>Steps required to submit a new application/workflow:</p>
<ul>
<li>
<p>Edit the <a href="https://github.com/kendomaniac/docker4seq/blob/devel/R/skeleton.R">skeleton.R</a> function and the ubuntu docker image (docker.io/repbioinfo/ubuntu) to create the new application.</p>
<ul>
<li>Please have a look at: <a href="https://kendomaniac.github.io/docker4seq/articles/skeleton.html">Controlling jobs in a docker image, a brief tutorial</a>.</li>
</ul>
</li>
<li><p>Create a public docker repository for the docker image, e.g. at <a href="http://docker.com">docker.com</a>.</p></li>
<li><p>Create a workflow.Rmd vignette using RStudio and publish it via RStudio. As example of a vignette see <a href="https://kendomaniac.github.io/docker4seq/">docker4seq vignette</a>.</p></li>
<li>
<p>Once the docker image, the function(s) and vignette are ready please fill this <a href="http://goo.gl/bb42EN">submission form</a>.</p>
<ul>
<li><p>If your module/workflow passes the validation analysis, you will be inserted as developer of the docker4seq package, and you will add your module and vignette to the docker4seq github repository.</p></li>
<li><p>If your module will not satisfy some of the validation points we will help you in fixing the issues and having your module compliant to the requirements of the Reproducible-bioinformatics Project. Mantainers will be responsable of the maintainance of their application(s).</p></li>
</ul>
</li>
</ul>
</div>
<div id="docker4seq" class="section level3">
<h3 class="hasAnchor">
<a href="#docker4seq" class="anchor"></a>docker4seq</h3>
<p>A collection of functions to execute NGS computing demanding applications, e.g reads mapping and counting, wrapped in docker containers. To install it you can use use <strong>devtools</strong>:</p>
<pre><code><a href="https://www.rdocumentation.org/packages/utils/topics/install.packages">install.packages("devtools")
library(devtools)
install_github("kendomaniac/docker4seq", ref="master")</a></code></pre>
<div id="requirements" class="section level4">
<h4 class="hasAnchor">
<a href="#requirements" class="anchor"></a>Requirements</h4>
<p>You need to have docker installed on your machine, for more info see this document: <a href="https://docs.docker.com/engine/installation/" class="uri">https://docs.docker.com/engine/installation/</a>. <strong>docker4seq</strong> package is expected to run on 64 bits linux machine with at least 4 cores. 32 Gb RAM are required only if mapping will be done with STAR. In case mapping is done with Salmon, only 16 Gb RAM are needed. A scratch folder should be present, e.g. /data/scratch and it should be writable from everybody:</p>
<pre><code>chmod 777 /data/scratch</code></pre>
<p>The functions in docker4seq package require that user is sudo or part of a docker group. See the following document for more info: <a href="https://docs.docker.com/install/linux/linux-postinstall/" class="uri">https://docs.docker.com/install/linux/linux-postinstall/</a></p>
<p><strong>IMPORTANT</strong> The first time <em>docker4seq</em> is installed the <strong>downloadContainers</strong> needs to be executed to download to the local repository the containers that are needed for the use of <em>docker4seq</em></p>
<p>More info on the functionalities of the package are available at: <a href="https://kendomaniac.github.io/docker4seq/index.html"><strong>docker4seq/4SeqGUI vignette</strong></a></p>
<ul>
<li>
<p><strong>docker4seq/4SeqGUI Video Tutorials:</strong></p>
<ul>
<li><p><a href="https://www.youtube.com/playlist?list=PLN48SoNXrLRhTi9MYMysNI3O4fR0wt46D">HowTo run a full RNAseq analysis</a></p></li>
<li><p><a href="https://www.youtube.com/playlist?list=PLN48SoNXrLRix-Er5unoze68qE8vW46s-">HowTo run a full miRNAseq analysis</a></p></li>
<li><p><a href="https://www.youtube.com/playlist?list=PLN48SoNXrLRhqjyPBGkDRTf0wNi4Y9Va2">HowTo run a full ChIPseq analysis</a></p></li>
</ul>
</li>
</ul>
<p><strong>testSeqbox</strong> In <em>docker4seq</em> library is now present the function <em>testSeqbox</em>, allowing to check if the software required for docker4seq functionalities is properly installed. Check <em>?testSeqbox</em> to see how to use it.</p>
</div>
<div id="workflows-compliance-with-sandve-rules" class="section level4">
<h4 class="hasAnchor">
<a href="#workflows-compliance-with-sandve-rules" class="anchor"></a>Workflows compliance with Sandve rules:</h4>
<ul>
<li><p>The <strong>whole transcriptome workflow</strong>, embedding annotatingByGtf.R, demultiplexing.R, experimentPower.R, fastqc.R, filterCounts.R, pca.R, rnaseqCounts.R, rsemAnnotate.R, rsemStar.R, rsemstarIndex.R, salmonAnnotation.R, salmonCounts.R, salmonIndex.R, samples2experiment.R, sampleSize.R, skewer.R, wrapperDeseq2.R, wrapperSalmon.R fullfils all <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003285">Sandve</a> reproducibility rules.</p></li>
<li><p>The <strong>miRNAs analysis workflow</strong>, embedding demultiplexing.R, experimentPower.R, fastqc.R, filterCounts.R, pca.R, mirnaCounts.R, mirnaCovar.R, sampleSize.R, wrapperDeseq2.R fullfils all <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003285">Sandve</a> reproducibility rules.</p></li>
<li><p>The <strong>Xenome module</strong>, which allows to discriminate between mouse and human reads in patient derived xenograft DNA/RNA sequenced samples, embedding xenome.R and xenomeIndex.R, fullfils all <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003285">Sandve</a> reproducibility rules.</p></li>
<li><p>The <strong>inDrop single cell module</strong>, which allows the single cell UMI counting generated with inDrop single cell sequencing technology, embedding indropCounts.R and demultiplexing.R fullfils all <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003285">Sandve</a> reproducibility rules.</p></li>
<li><p>The <strong>Platypus module</strong>, which allows haplotype-based variant calling for next generation sequence data, embedding demultiplexing.R, bwa.R and platypus.R fullfils all <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003285">Sandve</a></p></li>
<li><p>The <strong>Circular RNA identification module</strong>, which allows the identification circRNAs, embedding starChipIndex.R, starChimeric.R and starchipCircle.R fullfils all <a href="http://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1003285">Sandve</a></p></li>
<li><p>The <strong>ChIPseq workflow</strong> embedding demultiplexing.R, chipseq.R, chipseqCounts.R, bwa.R, bwaIndexUcsc.R does not satisfy Rule 4 (Version Control All Custom Scripts) because it download for annotation the latest version of ENSEMBL annotation. Thus, annotation executed at different time might differ because of the changes in the ENSEMBL downoaded information. We are working to fix this issue, expected fixing Q3 2018. Not all intermediate results are available as part of the final results (Sandve rule 5), expected fixing Q3 2018.</p></li>
</ul>
</div>
<div id="diclaimer" class="section level4">
<h4 class="hasAnchor">
<a href="#diclaimer" class="anchor"></a>Diclaimer:</h4>
<p>docker4seq developers have no liability for any use of docker4seq functions, including without limitation, any loss of data, incorrect results, or any costs, liabilities, or damages that result from use of docker4seq.</p>
</div>
</div>
<div id="seqgui-project" class="section level3">
<h3 class="hasAnchor">
<a href="#seqgui-project" class="anchor"></a>4SeqGUI Project</h3>
<p><a href="https://github.com/mbeccuti/4SeqGUI">4SeqGUI</a> is the GUI that can control the docker4seq functionalities. It represents the graphical interface used in SeqBox project (see below).</p>
<p>Video tutorials for 4SeqGUI:</p>
<p><a href="https://www.youtube.com/playlist?list=PLN48SoNXrLRhTi9MYMysNI3O4fR0wt46D">HowTo run a full RNAseq analysis</a></p>
<p><a href="https://www.youtube.com/playlist?list=PLN48SoNXrLRix-Er5unoze68qE8vW46s-">HowTo run a full miRNAseq analysis</a></p>
<p><a href="https://www.youtube.com/playlist?list=PLN48SoNXrLRhqjyPBGkDRTf0wNi4Y9Va2">HowTo run a full ChIPseq analysis</a></p>
</div>
<div id="the-seqbox-project" class="section level3">
<h3 class="hasAnchor">
<a href="#the-seqbox-project" class="anchor"></a>The SeqBox Project</h3>
<p>Short reads sequencing technology has been used for more than a decade now. However, the analysis of RNAseq and ChIPseq data is still computational demanding and the simple access to raw data does not guarantee results reproducibility between laboratories. To address these two aspects, we developed SeqBox, a cheap, efficient and reproducible RNAseq/ChIPseq hardware/software solution based on NUC6I7KYK mini-PC (an Intel consumer game computer with a fast processor and a high performance SSD disk), and Docker container platform. In SeqBox the analysis of RNAseq and ChIPseq data is supported by a friendly GUI. This allows access to fast and reproducible analyses also to scientists with/without scripting experience.</p>
<p>More info on SeqBox characteristics and cost are available at <a href="http://www.seqbox.com">www.seqbox.com</a></p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li>GPL (&gt;= 2)</li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>Raffaele Calogero <br><small class="roles"> Author, maintainer </small>  </li>
<li>Neha Kulkarni <br><small class="roles"> Author </small>  </li>
<li>Luca Alessandri <br><small class="roles"> Author </small>  </li>
<li>Marco Beccuti <br><small class="roles"> Author </small>  </li>
<li>Greta Romano <br><small class="roles"> Author </small>  </li>
<li>Francesca Cordero <br><small class="roles"> Author </small>  </li>
</ul>
</div>

  </div>
</div>

      <footer><div class="copyright">
  <p>Developed by Raffaele Calogero, Neha Kulkarni, Luca Alessandri, Marco Beccuti, Greta Romano, Francesca Cordero.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
