<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Detection and annotation of TF binding sites and histone marks, based on the workflow used in Galli et al Mol Cell. 2015 Oct 15;60(2):328-37 — chipseq • docker4seq</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Detection and annotation of TF binding sites and histone marks, based on the workflow used in Galli et al Mol Cell. 2015 Oct 15;60(2):328-37 — chipseq" />

<meta property="og:description" content="This function executes the docker container chipseq.8 and requires as input two bam files, one for the chipseq of interest and the other for control, e.g. mock, which can be generated with bwa function." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">docker4seq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.2.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/docker4seq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/skeleton.html">Skeleton tutorial</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Detection and annotation of TF binding sites and histone marks, based on the workflow used in Galli et al Mol Cell. 2015 Oct 15;60(2):328-37</h1>
    
    <div class="hidden name"><code>chipseq.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>This function executes the docker container chipseq.8 and requires as input two bam files, one for the chipseq of interest and the other for control, e.g. mock, which can be generated with bwa function.</p>
    
    </div>

    <pre class="usage"><span class='fu'>chipseq</span>(<span class='kw'>group</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"sudo"</span>, <span class='st'>"docker"</span>), <span class='kw'>bam.folder</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/getwd'>getwd</a></span>(), <span class='no'>sample.bam</span>,
  <span class='no'>ctrl.bam</span>, <span class='kw'>scratch.folder</span> <span class='kw'>=</span> <span class='st'>"/data/scratch"</span>, <span class='kw'>genome</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"hg19"</span>,
  <span class='st'>"hg38"</span>, <span class='st'>"mm9"</span>, <span class='st'>"mm10"</span>), <span class='no'>read.size</span>, <span class='kw'>tool</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"macs"</span>, <span class='st'>"sicer"</span>),
  <span class='kw'>macs.min.mfold</span> <span class='kw'>=</span> <span class='fl'>10</span>, <span class='kw'>macs.max.mfold</span> <span class='kw'>=</span> <span class='fl'>30</span>, <span class='kw'>macs.pval</span> <span class='kw'>=</span> <span class='st'>"1e-5"</span>,
  <span class='kw'>sicer.wsize</span> <span class='kw'>=</span> <span class='fl'>200</span>, <span class='kw'>sicer.gsize</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='fl'>200</span>, <span class='fl'>600</span>), <span class='kw'>sicer.fdr</span> <span class='kw'>=</span> <span class='fl'>0.1</span>,
  <span class='kw'>tss.distance</span> <span class='kw'>=</span> <span class='fl'>0</span>, <span class='kw'>max.upstream.distance</span> <span class='kw'>=</span> <span class='fl'>10000</span>,
  <span class='kw'>remove.duplicates</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"Y"</span>, <span class='st'>"N"</span>))</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>group, </th>
      <td><p>a character string. Two options: <code>"sudo"</code> or <code>"docker"</code>, depending to which group the user belongs</p></td>
    </tr>
    <tr>
      <th>bam.folder, </th>
      <td><p>a character string indicating where bam files are located</p></td>
    </tr>
    <tr>
      <th>sample.bam, </th>
      <td><p>a character string indicating the chipseq file under analysis</p></td>
    </tr>
    <tr>
      <th>ctrl.bam, </th>
      <td><p>a character string indicating the control file, e.g. unspecific IgG, input DNA, etc.</p></td>
    </tr>
    <tr>
      <th>scratch.folder, </th>
      <td><p>a character string indicating the scratch folder where docker container will be mounted</p></td>
    </tr>
    <tr>
      <th>genome, </th>
      <td><p>a character string indicating the genome used as reference for data generation. Available options: hg19, hg38, mm9, mm10</p></td>
    </tr>
    <tr>
      <th>read.size, </th>
      <td><p>an integer indicating the length of the sequenced reads</p></td>
    </tr>
    <tr>
      <th>tool, </th>
      <td><p>a character string indicating the peaks calling algorith. Available options: macs and sicer. Macs, v 1.14, is used to call TF peaks, as instead sicer, v 1.1, is used to call histone mark peaks</p></td>
    </tr>
    <tr>
      <th>macs.min.mfold, </th>
      <td><p>an integer indicating the minimum enrichment ratio against background</p></td>
    </tr>
    <tr>
      <th>macs.max.mfold, </th>
      <td><p>an integer indicating the maximum enrichment ratio against background</p></td>
    </tr>
    <tr>
      <th>macs.pval, </th>
      <td><p>a character string, indicationg the pvalue cutoff to be used to filter peaks with low statistical significance.The number must be provided in scientific notation as the default value shows</p></td>
    </tr>
    <tr>
      <th>sicer.wsize, </th>
      <td><p>an integer indicating the windows size to be used by sicer</p></td>
    </tr>
    <tr>
      <th>sicer.gsize, </th>
      <td><p>an integer indicating the gap size to be used by sicer. Suggested values: H3K4Me3=200; H3K27Me3=600</p></td>
    </tr>
    <tr>
      <th>sicer.fdr, </th>
      <td><p>an integer indicating the pvalue cutoff to be used to filter peaks with low statistical significance</p></td>
    </tr>
    <tr>
      <th>tss.distance, </th>
      <td><p>an integer indicating the distance of TSS with respect to gene start</p></td>
    </tr>
    <tr>
      <th>max.upstream.distance, </th>
      <td><p>an integer indicating the maximum distance to associate a gene ID to a peak</p></td>
    </tr>
    <tr>
      <th>remove.duplicates, </th>
      <td><p>a character string indicating if duplicated reads have to be removed. Available options: Y, to remove douplicates, N to keep duplicates</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>three files: dedup_reads.bam, which is sorted and duplicates marked bam file, dedup_reads.bai, which is the index of the dedup_reads.bam, and dedup_reads.stats, which provides mapping statistics</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
    <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system'>system</a></span>(<span class='st'>"wget http://130.192.119.59/public/SRR1172111.bam"</span>)<span class='co'>#TEAD</span>
    <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system'>system</a></span>(<span class='st'>"wget http://130.192.119.59/public/SRR1172110.bam"</span>)<span class='co'>#igg</span>
    <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system'>system</a></span>(<span class='st'>"wget http://130.192.119.59/public/SRR1592211.bam"</span>)<span class='co'>#H3K27ac</span>
    <span class='co'>#running chipseq for macs</span>
    <span class='fu'>chipseq</span>(<span class='kw'>group</span><span class='kw'>=</span><span class='st'>"sudo"</span>,<span class='kw'>bam.folder</span><span class='kw'>=</span><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/getwd'>getwd</a></span>(), <span class='kw'>sample.bam</span><span class='kw'>=</span><span class='st'>"SRR1172111.bam"</span>, <span class='kw'>ctrl.bam</span><span class='kw'>=</span><span class='st'>"SRR1172110.bam"</span>,
    <span class='kw'>scratch.folder</span><span class='kw'>=</span><span class='st'>"/data/scratch"</span>, <span class='kw'>genome</span><span class='kw'>=</span><span class='st'>"hg19"</span>, <span class='kw'>read.size</span><span class='kw'>=</span><span class='fl'>50</span>,
    <span class='kw'>tool</span><span class='kw'>=</span><span class='st'>"macs"</span>, <span class='kw'>macs.min.mfold</span><span class='kw'>=</span><span class='fl'>10</span>, <span class='kw'>macs.max.mfold</span><span class='kw'>=</span><span class='fl'>30</span>, <span class='kw'>macs.pval</span><span class='kw'>=</span><span class='st'>"1e-5"</span>,
    <span class='kw'>sicer.wsize</span><span class='kw'>=</span><span class='fl'>200</span>, <span class='kw'>sicer.gsize</span><span class='kw'>=</span><span class='fl'>200</span>, <span class='kw'>sicer.fdr</span><span class='kw'>=</span><span class='fl'>0.10</span>, <span class='kw'>tss.distance</span><span class='kw'>=</span><span class='fl'>0</span>, <span class='kw'>max.upstream.distance</span><span class='kw'>=</span><span class='fl'>10000</span>,
    <span class='kw'>remove.duplicates</span><span class='kw'>=</span><span class='st'>"N"</span>)

    <span class='co'>#running chipseq for sicer H3K4Me3</span>
    <span class='fu'>chipseq</span>(<span class='kw'>group</span><span class='kw'>=</span><span class='st'>"sudo"</span>,<span class='kw'>bam.folder</span><span class='kw'>=</span><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/getwd'>getwd</a></span>(), <span class='kw'>sample.bam</span><span class='kw'>=</span><span class='st'>"SRR1592211.bam"</span>, <span class='kw'>ctrl.bam</span><span class='kw'>=</span><span class='st'>"SRR1172110.bam"</span>,
    <span class='kw'>scratch.folder</span><span class='kw'>=</span><span class='st'>"/data/scratch"</span>, <span class='kw'>genome</span><span class='kw'>=</span><span class='st'>"hg19"</span>, <span class='kw'>read.size</span><span class='kw'>=</span><span class='fl'>50</span>,
    <span class='kw'>tool</span><span class='kw'>=</span><span class='st'>"sicer"</span>, <span class='kw'>sicer.wsize</span><span class='kw'>=</span><span class='fl'>200</span>, <span class='kw'>sicer.gsize</span><span class='kw'>=</span><span class='fl'>200</span>, <span class='kw'>sicer.fdr</span><span class='kw'>=</span><span class='fl'>0.10</span>,
    <span class='kw'>tss.distance</span><span class='kw'>=</span><span class='fl'>0</span>, <span class='kw'>max.upstream.distance</span><span class='kw'>=</span><span class='fl'>10000</span>,<span class='kw'>remove.duplicates</span><span class='kw'>=</span><span class='st'>"N"</span>)
<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Raffaele Calogero, Neha Kulkarni, Luca Alessandri, Marco Beccuti, Greta Romano, Francesca Cordero.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

