<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Running RNAseq counting workflow for a single sample — rnaseqCounts • docker4seq</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="Running RNAseq counting workflow for a single sample — rnaseqCounts" />

<meta property="og:description" content="This function executes a set of docker containers allowing the generation of gene and isoforms counts for a single sample.
#params skewer" />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">docker4seq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.2.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/docker4seq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/skeleton.html">Skeleton tutorial</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Running RNAseq counting workflow for a single sample</h1>
    
    <div class="hidden name"><code>rnaseqCounts.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>This function executes a set of docker containers allowing the generation of gene and isoforms counts for a single sample.
#params skewer</p>
    
    </div>

    <pre class="usage"><span class='fu'>rnaseqCounts</span>(<span class='kw'>group</span> <span class='kw'>=</span> <span class='st'>"sudo"</span>, <span class='kw'>fastq.folder</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/getwd'>getwd</a></span>(),
  <span class='kw'>scratch.folder</span> <span class='kw'>=</span> <span class='st'>"/data/scratch"</span>, <span class='kw'>threads</span> <span class='kw'>=</span> <span class='fl'>4</span>, <span class='no'>adapter5</span>, <span class='no'>adapter3</span>,
  <span class='kw'>seq.type</span> <span class='kw'>=</span> <span class='st'>"pe"</span>, <span class='kw'>min.length</span> <span class='kw'>=</span> <span class='fl'>40</span>,
  <span class='kw'>genome.folder</span> <span class='kw'>=</span> <span class='st'>"/data/genomes/hg38star"</span>, <span class='kw'>strandness</span> <span class='kw'>=</span> <span class='st'>"none"</span>,
  <span class='kw'>save.bam</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>org</span> <span class='kw'>=</span> <span class='st'>"hg38"</span>, <span class='kw'>annotation.type</span> <span class='kw'>=</span> <span class='st'>"gtfENSEMBL"</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>group, </th>
      <td><p>a character string. Two options: <code>"sudo"</code> or <code>"docker"</code>, depending to which group the user belongs</p></td>
    </tr>
    <tr>
      <th>fastq.folder, </th>
      <td><p>a character string indicating where gzip fastq files are located</p></td>
    </tr>
    <tr>
      <th>scratch.folder, </th>
      <td><p>a character string indicating the scratch folder where docker container will be mounted</p></td>
    </tr>
    <tr>
      <th>threads, </th>
      <td><p>a number indicating the number of cores to be used from the application</p></td>
    </tr>
    <tr>
      <th>adapter5, </th>
      <td><p>a character string indicating the fwd adapter</p></td>
    </tr>
    <tr>
      <th>adapter3, </th>
      <td><p>a character string indicating the rev adapter</p></td>
    </tr>
    <tr>
      <th>seq.type, </th>
      <td><p>a character string indicating the type of reads to be generated by the sequencer. Two options: <code>"se"</code> or <code>"pe"</code> respectively for single end and pair end sequencing.</p></td>
    </tr>
    <tr>
      <th>min.length, </th>
      <td><p>a number indicating minimal length required to return a trimmed read
#params rsemstar</p></td>
    </tr>
    <tr>
      <th>genome.folder, </th>
      <td><p>a character string indicating the folder where the indexed reference genome for bwa is located. IMPORTANT the present function only suport genomic indexes made using ensembl genom and the corresponding gtf</p></td>
    </tr>
    <tr>
      <th>strandness, </th>
      <td><p>a character string indicating the type ofsequencing protocol used for the analysis. Three options: <code>"none"</code>, <code>"forward"</code>, <code>"reverse"</code> respectively for non strand selection, reverse for Illumina strandness protocols, reverse for ACCESS Illumina protocol</p></td>
    </tr>
    <tr>
      <th>save.bam, </th>
      <td><p>a boolean value, TRUE or FALSE, to save also BAM files generated by STAR and RSEM
#params rsemanno</p></td>
    </tr>
    <tr>
      <th>org, </th>
      <td><p>a character string indicating the genome assembly used for mapping and counting with <code>"rsemstar"</code> function only required for biocENSEMBL based annotation</p></td>
    </tr>
    <tr>
      <th>annotation.type, </th>
      <td><p>a character string. Two options: <code>"biocENSEMBL"</code> or <code>"gtfENSEMBL"</code>. <code>"biocENSEMBL"</code> will annotate by Bioconductor only protein coding genes. <code>"gtfENSEMBL"</code> will annotate all RNAs described in <code>"annotation.type"</code></p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Returns the output of skewer, rsemstar, rsemannos' functions</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system'>system</a></span>(<span class='st'>"wget http://130.192.119.59/public/test_R1.fastq.gz"</span>)
<span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/library'>library</a></span>(<span class='no'>docker4seq</span>)
<span class='fu'>rnaseqCounts</span>(<span class='kw'>group</span><span class='kw'>=</span><span class='st'>"docker"</span>,<span class='kw'>fastq.folder</span><span class='kw'>=</span><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/getwd'>getwd</a></span>(), <span class='kw'>scratch.folder</span><span class='kw'>=</span><span class='st'>"/data/scratch/"</span>,
            <span class='kw'>adapter5</span><span class='kw'>=</span><span class='st'>"AGATCGGAAGAGCACACGTCTGAACTCCAGTCA"</span>,
            <span class='kw'>adapter3</span><span class='kw'>=</span><span class='st'>"AGATCGGAAGAGCGTCGTGTAGGGAAAGAGTGT"</span>,
            <span class='kw'>seq.type</span><span class='kw'>=</span><span class='st'>"se"</span>, <span class='kw'>threads</span><span class='kw'>=</span><span class='fl'>24</span>,  <span class='kw'>min.length</span><span class='kw'>=</span><span class='fl'>40</span>,
            <span class='kw'>genome.folder</span><span class='kw'>=</span><span class='st'>"/data/genomes/hg38star"</span>, <span class='kw'>strandness</span><span class='kw'>=</span><span class='st'>"none"</span>, <span class='kw'>save.bam</span><span class='kw'>=</span><span class='fl'>FALSE</span>,
            <span class='kw'>org</span><span class='kw'>=</span><span class='st'>"hg38"</span>, <span class='kw'>annotation.type</span><span class='kw'>=</span><span class='st'>"gtfENSEMBL"</span>)
<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Raffaele Calogero</p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Raffaele Calogero, Neha Kulkarni, Luca Alessandri, Marco Beccuti, Greta Romano, Francesca Cordero.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

