<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>generating counts, FPKM and TPM tables from rnaseqCounts outuputs — sample2experiment • docker4seq</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



<meta property="og:title" content="generating counts, FPKM and TPM tables from rnaseqCounts outuputs — sample2experiment" />

<meta property="og:description" content="This function generates counts, FPKM and TPM tables from rnaseqCounts outuputs." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">docker4seq</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.2.9</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../articles/docker4seq.html">Get started</a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
    <li>
      <a href="../articles/skeleton.html">Skeleton tutorial</a>
    </li>
  </ul>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>generating counts, FPKM and TPM tables from rnaseqCounts outuputs</h1>
    
    <div class="hidden name"><code>sample2experiment.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>This function generates counts, FPKM and TPM tables from rnaseqCounts outuputs.</p>
    
    </div>

    <pre class="usage"><span class='fu'>sample2experiment</span>(<span class='no'>sample.folders</span>, <span class='no'>covariates</span>, <span class='kw'>batch</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>bio.type</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"protein_coding"</span>, <span class='st'>"unitary_pseudogene"</span>,
  <span class='st'>"unprocessed_pseudogene"</span>, <span class='st'>"processed_pseudogene"</span>,
  <span class='st'>"transcribed_unprocessed_pseudogene"</span>, <span class='st'>"processed_transcript"</span>,
  <span class='st'>"antisense"</span>, <span class='st'>"transcribed_unitary_pseudogene"</span>, <span class='st'>"polymorphic_pseudogene"</span>,
  <span class='st'>"lincRNA"</span>, <span class='st'>"sense_intronic"</span>, <span class='st'>"transcribed_processed_pseudogene"</span>,
  <span class='st'>"sense_overlapping"</span>, <span class='st'>"IG_V_pseudogene"</span>, <span class='st'>"pseudogene"</span>, <span class='st'>"TR_V_gene"</span>,
  <span class='st'>"3prime_overlapping_ncRNA"</span>, <span class='st'>"IG_V_gene"</span>, <span class='st'>"bidirectional_promoter_lncRNA"</span>,
  <span class='st'>"snRNA"</span>, <span class='st'>"miRNA"</span>, <span class='st'>"misc_RNA"</span>, <span class='st'>"snoRNA"</span>, <span class='st'>"rRNA"</span>, <span class='st'>"IG_C_gene"</span>, <span class='st'>"IG_J_gene"</span>,
       <span class='st'>"TR_J_gene"</span>, <span class='st'>"TR_C_gene"</span>, <span class='st'>"TR_V_pseudogene"</span>, <span class='st'>"TR_J_pseudogene"</span>,
  <span class='st'>"IG_D_gene"</span>, <span class='st'>"ribozyme"</span>, <span class='st'>"IG_C_pseudogene"</span>, <span class='st'>"TR_D_gene"</span>, <span class='st'>"TEC"</span>,
  <span class='st'>"IG_J_pseudogene"</span>, <span class='st'>"scRNA"</span>, <span class='st'>"scaRNA"</span>, <span class='st'>"vaultRNA"</span>, <span class='st'>"sRNA"</span>, <span class='st'>"macro_lncRNA"</span>,
  <span class='st'>"non_coding"</span>, <span class='st'>"IG_pseudogene"</span>), <span class='kw'>output.prefix</span> <span class='kw'>=</span> <span class='st'>"."</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>sample.folders, </th>
      <td><p>a character string indicating the paths of rnaseqCouts output folders</p></td>
    </tr>
    <tr>
      <th>covariates, </th>
      <td><p>a character string indicating the covariates associated to each sample. Covariates are required for differnetial expression analysis</p></td>
    </tr>
    <tr>
      <th>batch, </th>
      <td><p>a character string indicating the batch associated to each sample</p></td>
    </tr>
    <tr>
      <th>bio.type, </th>
      <td><p>a character string indicating the ensemb bio.type. Options: "protein_coding","unitary_pseudogene","unprocessed_pseudogene","processed_pseudogene", "transcribed_unprocessed_pseudogene","processed_transcript","antisense","transcribed_unitary_pseudogene","polymorphic_pseudogene","lincRNA","sense_intronic","transcribed_processed_pseudogene","sense_overlapping","IG_V_pseudogene","pseudogene","TR_V_gene","3prime_overlapping_ncRNA","IG_V_gene","bidirectional_promoter_lncRNA","snRNA","miRNA","misc_RNA","snoRNA","rRNA","IG_C_gene","IG_J_gene","TR_J_gene","TR_C_gene","TR_V_pseudogene","TR_J_pseudogene","IG_D_gene","ribozyme","IG_C_pseudogene","TR_D_gene","TEC","IG_J_pseudogene","scRNA","scaRNA","vaultRNA","sRNA","macro_lncRNA","non_coding","IG_pseudogene"</p></td>
    </tr>
    <tr>
      <th>output.prefix, </th>
      <td><p>a character value indicating the output folder path</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Returns counts, fpkm, tpm data frames for gene and isoforms, save data frames in experiment.tables.Rda, in counts.txt, log2fpkm.txt and in log2TPM</p>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><span class='co'># NOT RUN {</span>
  <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system'>system</a></span>(<span class='st'>"wget http://130.192.119.59/public/test.samples2experiment.zip"</span>)
  <span class='fu'><a href='https://www.rdocumentation.org/packages/utils/topics/unzip'>unzip</a></span>(<span class='st'>"test.samples2experiment.zip"</span>)
  <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/getwd'>setwd</a></span>(<span class='st'>"test.samples2experiment"</span>)
  <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/library'>library</a></span>(<span class='no'>docker4seq</span>)
  <span class='fu'>sample2experiment</span>(<span class='kw'>sample.folders</span><span class='kw'>=</span><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"./e1g"</span>,<span class='st'>"./e2g"</span>,<span class='st'>"./e3g"</span>,
                 <span class='st'>"./p1g"</span>, <span class='st'>"./p2g"</span>, <span class='st'>"./p3g"</span>),
                 <span class='kw'>covariates</span><span class='kw'>=</span><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/c'>c</a></span>(<span class='st'>"Cov.1"</span>,<span class='st'>"Cov.1"</span>,<span class='st'>"Cov.1"</span>,
                 <span class='st'>"Cov.2"</span>,<span class='st'>"Cov.2"</span>,<span class='st'>"Cov.2"</span>),
                 <span class='kw'>bio.type</span><span class='kw'>=</span><span class='st'>"protein_coding"</span>, <span class='kw'>output.prefix</span><span class='kw'>=</span><span class='st'>"."</span>)
<span class='co'># }</span></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Raffaele Calogero</p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by Raffaele Calogero, Neha Kulkarni, Luca Alessandri, Marco Beccuti, Greta Romano, Francesca Cordero.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

